cdc/scheduler/internal/v3/compat/compat.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package compat
+
+import (
+	"github.com/coreos/go-semver/semver"
+	"github.com/pingcap/tiflow/cdc/model"
+	"github.com/pingcap/tiflow/cdc/processor/tablepb"
+	"github.com/pingcap/tiflow/cdc/scheduler/schedulepb"
+	"github.com/pingcap/tiflow/pkg/config"
+	"github.com/pingcap/tiflow/pkg/spanz"
+	"github.com/pingcap/tiflow/pkg/version"
+)
+
+// Compat is a compatibility layer between span replication and table replication.
+type Compat struct {
+	enableSpanReplication bool
+
+	hasChecked  bool
+	config      *config.SchedulerConfig
+	captureInfo map[model.CaptureID]*model.CaptureInfo
+}
+
+// New returns a new Compat.
+func New(
+	config *config.SchedulerConfig,
+	captureInfo map[model.CaptureID]*model.CaptureInfo,
+) *Compat {
+	return &Compat{
+		config:      config,
+		captureInfo: captureInfo,
+	}
+}
+
+// UpdateCaptureInfo update the latest alive capture info.
+// Returns true if capture info has changed.
+func (c *Compat) UpdateCaptureInfo(
+	aliveCaptures map[model.CaptureID]*model.CaptureInfo,
+) bool {
+	if len(aliveCaptures) != len(c.captureInfo) {
+		c.captureInfo = aliveCaptures
+		c.hasChecked = false
+		return true
+	}
+	for id := range aliveCaptures {
+		_, ok := c.captureInfo[id]
+		if !ok {
+			c.captureInfo = aliveCaptures
+			c.hasChecked = false
+			return true
+		}
+	}
+	return false
+}
+
+// SpanReplicationMinVersion is the min version that allows span replication.
+// For now, we ASSUME it is `v6.6.0-alpha`.
+// FIXME: correct the min version before we release the feature.
+var SpanReplicationMinVersion = semver.New("6.6.0-alpha")
+
+// CheckSpanReplicationEnabled check if the changefeed can enable span replication.
+func (c *Compat) CheckSpanReplicationEnabled() bool {
+	if c.hasChecked {
+		return c.enableSpanReplication
+	}
+	c.hasChecked = true
+
+	c.enableSpanReplication = c.config.RegionPerSpan != 0
```

Comment:
`config` is only used to here, so no need to keep it as a field of the `Compat`, replace it by `enableSpanReplication` when new the compat.